### PR TITLE
My PR Classic Frogger Game

### PR DESCRIPTION
**Pull request by Nik3264**
### Classic Frogger Game

[Demo](https://nik3264.github.io/frontend-nanodegree-arcade-game/) |
[Code base](https://github.com/Nik3264/frontend-nanodegree-arcade-game)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

**Comment by Nik3264** [comment]
I've looked at the code and everything seems to be working. Please, review.

**Comment by Nik3264** [comment]
I've looked at the code and everything seems to be working. Please, review.

**Comment by Nik3264** [comment]
Please, review.

**Comment by Nik3264** [comment]
I saw comments from other students and made changes. I removed the constants in the properties of the Game class.

File: submissions/Nik3264/Frogger-Game/app.js [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  update(dt) {
+    this.x +=
+      dt * (this.speed + 0.4 * this.speed * game.getLevel()) * !game.isStop;
+    if (this.x > game.frontierForEnemiesRight) {
+      this.x = game.frontierForEnemiesLeft;
+    }
+    if (this.isCollision()) {
+      player.start();
```
Comment (OleksiyRudenko) [inline submissions/Nik3264/Frogger-Game/app.js:76]: Item 4.iii of the requirements says:
> Classes do not refer to any global variables, like global variable player, which is an instance of Player class (referring to global constants and globals provided by the gaming platform like Resources is OK); Hint: pass instance of a game object (or objects) as an argument to other game objects they need to interact with.

It is relevant to this line of code and other methods where global variable `player` is referred to.

Same problem is with variable `game`. Besides, why do you need to refer to a specific instance of class `Game` within methods of that class? I mean you perfectly fine use `this` to refer to the properties of class `Game` and suddenly start using specific instance of the class.

File: submissions/Nik3264/Frogger-Game/app.js [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    if (
+      this.x > player.x - game.collisionX &&
+      this.x < player.x + game.collisionX &&
+      this.y > player.y - game.collisionY &&
+      this.y < player.y + game.collisionY
+    ) {
+      return true;
+    } else {
+      return false;
+    }
```
Comment (OleksiyRudenko) [inline submissions/Nik3264/Frogger-Game/app.js:91]: Redundant.
`if`'s condition resolves to either `true` or `false` and based on that `then` or `else` part of `if` is executed.
You may simply return the values testing condition here.

File: submissions/Nik3264/Frogger-Game/app.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  render() {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+  }
+}
+
+class Player {
+  constructor() {
+    this.sprite = "images/char-boy.png";
+    this.x = game.PLAYER_X_START;
+    this.y = game.PLAYER_Y_START;
```
Comment (OleksiyRudenko) [inline submissions/Nik3264/Frogger-Game/app.js:103]: Item 4.iii of the requirements. Here and everywhere in classes definitions across the code.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@Nik3264 good progress.
A few issues yet to solve.

File: submissions/Nik3264/Frogger-Game/app.js [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        //console.log(this.isCollision);
+        if (this.isCollision()){
+            player.start();
+            game.setLevel();
+        }
+    }
+
+    isCollision(){
+        if (  (this.x > player.x-50 && this.x < player.x+50) &&
+              (this.y > player.y-40 && this.y < player.y+40) ) {
```
[the remark this answers is not in the document]
Comment (Nik3264) [reply submissions/Nik3264/Frogger-Game/app.js]: i fixed it

File: submissions/Nik3264/Frogger-Game/app.js [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+    handleInput(key){
+        switch(key){
+            case "up":
+                if(this.y>0){
+                    this.y-=CELL_HEIGHT; 
+                }                
+                break;
+            case "down":
+                if(this.y<383){
```
[the remark this answers is not in the document]
Comment (Nik3264) [reply submissions/Nik3264/Frogger-Game/app.js]: i fixed it

File: submissions/Nik3264/Frogger-Game/app.js (hunk 1 above)
Comment (Nik3264) [reply submissions/Nik3264/Frogger-Game/app.js]: I added the player object as a parameter

File: submissions/Nik3264/Frogger-Game/app.js (hunk 2 above)
Comment (Nik3264) [reply submissions/Nik3264/Frogger-Game/app.js]: it was so obvious, but I didn't notice it! Thank you

File: submissions/Nik3264/Frogger-Game/app.js (hunk 3 above)
Comment (Nik3264) [reply submissions/Nik3264/Frogger-Game/app.js]: done

File: submissions/Nik3264/Frogger-Game/app.js [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+      body.innerHTML = "";
+    }, 1500);
+  }
+
+  isWin() {
+    if (this.y <= 0) {
+      return true;
+    } else {
+      return false;
+    }
```
[the remark this answers is not in the document]
Comment (Nik3264) [reply submissions/Nik3264/Frogger-Game/app.js]: done

File: submissions/Nik3264/Frogger-Game/app.js [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+  }
+  render() {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+  }
+}
+
+const game = new Game(gameProperties);
+const bug1 = new Enemy({ x: 0, y: 140, speed: 100 });
+const bug2 = new Enemy({ x: 0, y: 50, speed: 50 });
+const bug3 = new Enemy({ x: 0, y: 230, speed: 200 });
```
[the remark this answers is not in the document]
Comment (Nik3264) [reply submissions/Nik3264/Frogger-Game/app.js]: I liked this part the most! ))
I tried to fix everything according to the requirements
Thank you.

File: submissions/Nik3264/Frogger-Game/app.js [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+const game = new Game(gameProperties);
+const player = new Player(game);
+
+const createEnemy = (numberOfEnemies) => {
+  const arrOfEnemies = [];
+  for (let i = 0; i < numberOfEnemies; i++) {
+    arrOfEnemies.push(i);
+  }
+  return arrOfEnemies;
```
Comment (OleksiyRudenko) [inline submissions/Nik3264/Frogger-Game/app.js:177]: Bonus: one of the methods to create an array knowing its length and pre-fill it with some values based on index is
`Array.from({length: numberOfEnemies}, (_, index) => index)`
Prepend with `return` and call it a day.
Still worth having a function for this one-liner as function provides a semantic meaning to the action.

**Comment by Nik3264** [comment]
I added inheritance and got rid of magic numbers. 
Thank you a lot

File: submissions/Nik3264/Frogger-Game/app.js [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+};
+
+const allEnemies = createEnemy(3).map((number) => {
+  return new Enemy({
+    x: 0,
+    y: game.startYForCreateEnemy + number * game.stepYForCreateEnemy,
+    speed: game.increaseSpeed + number * game.increaseSpeed,
+    game: game,
+    player: player,
```
Comment (OleksiyRudenko) [inline submissions/Nik3264/Frogger-Game/app.js:186]: There is legal shortcut to create object properties from variables where key will have the name same as the variable in use.
```
{
  variableName,
}
// ..is equivalent to...
{
  variableName: variableName,
}

```

**Review by OleksiyRudenko — APPROVED** [review]
@Nik3264 great job done!

**Pull request merged**